allennlp/models/encoder_decoders/composed_seq2seq.py

```diff
@@ -0,0 +1,130 @@
+from allennlp.nn import util, InitializerApplicator, RegularizerApplicator
+
+
+@Model.register("composed_seq2seq")
+class ComposedSeq2Seq(Model):
+    """
+    This ``ComposedSeq2Seq`` class is a :class:`Model` which takes a sequence, encodes it, and then
+    uses the encoded representations to decode another sequence.  You can use this as the basis for
+    a neural machine translation system, an abstractive summarization system, or any other common
+    seq2seq problem.  The model here is simple, but should be a decent starting place for
+    implementing recent models for these tasks.
+    The ``ComposedSeq2Seq`` is composed from separate Seq2SeqEncoder and DecoderCell classes.
+    This parts are customizable and independent from each other.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``, required
+        Vocabulary containing source and target vocabularies. They may be under the same namespace
+        (`tokens`) or the target tokens can have a different namespace, in which case it needs to
+        be specified as `target_namespace`.
+    source_text_embedders : ``TextFieldEmbedder``, required
+        Embedders for source side sequences
+    encoder : ``Seq2SeqEncoder``, required
+        The encoder of the "encoder/decoder" model
+    decoder : ``SeqDecoder``, required
+        The decoder of the "encoder/decoder" model
+    tied_source_embedder_key : ``str``, optional (default=``None``)
+        If specified, this key is used to obtain token_embedder in `source_text_embedder` and
+        the weights are shared/tied with the decoder's target embedding weights.
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+
+    def __init__(self,
+                 vocab: Vocabulary,
+                 source_text_embedder: TextFieldEmbedder,
+                 encoder: Seq2SeqEncoder,
+                 decoder: SeqDecoder,
+                 tied_source_embedder_key: Optional[str] = None,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None)-> None:
+
+        super(ComposedSeq2Seq, self).__init__(vocab, regularizer)
+
+        self._source_text_embedder = source_text_embedder
+        self._encoder = encoder
+        self._decoder = decoder
+
+        if self._encoder.get_output_dim() != self._decoder.get_output_dim():
+            raise ConfigurationError(f"Encoder output dimension {self._encoder.get_output_dim()} should be"
+                                     f" equal to decoder dimension {self._decoder.get_output_dim()}.")
+        if tied_source_embedder_key:
+            # Very ugly way to tie Embeddings
```

Comment:
Yeah, this isn't ideal, but it seems worth it regardless. Could you explain this limitation in the comment above for `tied_source_embedder_key`?

Reply:
@brendan-ai2 Can we in future explore whether tying weights with something like Initializor's regex selection be possible, or is it too niche and generally not applicable?

Reply:
Weight tying via config does sound interesting, but I can't claim much insight into how generally useful it would be. I'd defer to other maintainers on that. But we can certainly discuss it in another PR or issue! :)